florist/api/routes/server/training.py

```diff
@@ -126,30 +129,31 @@ async def start(job_id: str, request: Request, background_tasks: BackgroundTasks
         return JSONResponse({"error": str(ex)}, status_code=500)
 
 
-def client_training_listener(job: Job, client_info: ClientInfo, database: Database[Dict[str, Any]]) -> None:
+async def client_training_listener(job: Job, client_info: ClientInfo) -> None:
     """
     Listen to the Redis' channel that reports updates on the training process of a FL client.
 
     Keeps consuming updates to the channel until it finds `shutdown` in the client metrics.
 
     :param job: (Job) The job that has this client's metrics.
     :param client_info: (ClientInfo) The ClientInfo with the client_uuid to listen to.
-    :param database: (pymongo.database.Database) An instance of the database to save the information
-        into the Job. MUST BE A SYNCHRONOUS DATABASE since this function cannot be marked as async
-        because of limitations with FastAPI's BackgroundTasks.
     """
     LOGGER.info(f"Starting listener for client messages from job {job.id} at channel {client_info.uuid}")
 
     assert client_info.uuid is not None, "client_info.uuid is None."
 
+    db_client: AsyncIOMotorClient[Any] = AsyncIOMotorClient(MONGODB_URI)
+    database = db_client[DATABASE_NAME]
```

Comment:
This is sort of an interesting structure that I'm not super familiar with. For my own curiosity, what is happening when you do `db_client[DATABASE_NAME]`? I've only ever seen [...] for generics in mypy

Reply:
Ah...this was probably a dumb question. I assume this is something like a map that retrieves the database from the key 🤦

Reply:
This is an interesting thing, I went a bit deep into that because I had to figure out how to mock it for unit tests haha. It is not a map/dictionary, it's actually a class that implements the `__getitem__` function, which makes it also behave like a dictionary. I guess it's a bit cleaner to do that than to implement a `get_database` kind of method but it makes it a bit confusing at first. At least I learned something :)

See: https://github.com/VectorInstitute/FLorist/pull/122/files#diff-a35d7b576c111704ed16fa1a1c1fa155e1bf0ab6e5e7cbc3c34b22324cb7b959R625

Reply:
Gotcha...very similar, but much less obvious what's happening 😂